images/controller/internal/controllers/rvr_access_count/reconciler_test.go

```diff
@@ -0,0 +1,589 @@
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+	"sigs.k8s.io/controller-runtime/pkg/client/interceptor"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/deckhouse/sds-replicated-volume/api/v1alpha1"
+	"github.com/deckhouse/sds-replicated-volume/api/v1alpha3"
+	rvraccesscount "github.com/deckhouse/sds-replicated-volume/images/controller/internal/controllers/rvr_access_count"
+)
+
+var _ = Describe("Reconciler", func() {
+	var (
+		clientBuilder *fake.ClientBuilder
+		scheme        *runtime.Scheme
+		cl            client.WithWatch
+		rec           *rvraccesscount.Reconciler
+	)
+
+	BeforeEach(func() {
+		scheme = runtime.NewScheme()
+		Expect(v1alpha3.AddToScheme(scheme)).To(Succeed(), "should add v1alpha3 to scheme")
+		Expect(v1alpha1.AddToScheme(scheme)).To(Succeed(), "should add v1alpha1 to scheme")
+		clientBuilder = fake.NewClientBuilder().
+			WithScheme(scheme).
+			WithStatusSubresource(&v1alpha3.ReplicatedVolume{}, &v1alpha3.ReplicatedVolumeReplica{})
+	})
+
+	JustBeforeEach(func() {
+		cl = clientBuilder.Build()
+		rec = rvraccesscount.NewReconciler(cl, GinkgoLogr, scheme)
+	})
+
+	It("returns no error when ReplicatedVolume does not exist", func(ctx SpecContext) {
+		Expect(rec.Reconcile(ctx, reconcile.Request{
+			NamespacedName: types.NamespacedName{Name: "non-existent"},
+		})).ToNot(Requeue(), "should ignore NotFound errors")
+	})
+
+	When("Get RV fails with non-NotFound error", func() {
+		var testError error
+
+		BeforeEach(func() {
+			testError = errors.New("internal server error")
+			clientBuilder = clientBuilder.WithInterceptorFuncs(
+				InterceptGet(func(_ *v1alpha3.ReplicatedVolume) error {
+					return testError
+				}),
+			)
+		})
+
+		It("should return error", func(ctx SpecContext) {
+			Expect(rec.Reconcile(ctx, reconcile.Request{
+				NamespacedName: types.NamespacedName{Name: "test-rv"},
+			})).Error().To(MatchError(testError), "should return error when Get fails")
+		})
+	})
+
+	When("RV created", func() {
+		var (
+			rv  *v1alpha3.ReplicatedVolume
+			rsc *v1alpha1.ReplicatedStorageClass
+		)
+
+		BeforeEach(func() {
+			rv = &v1alpha3.ReplicatedVolume{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "test-volume",
+					UID:  "test-uid",
+				},
+				Spec: v1alpha3.ReplicatedVolumeSpec{
+					ReplicatedStorageClassName: "test-rsc",
+					PublishOn:                  []string{},
+				},
+			}
+			rsc = &v1alpha1.ReplicatedStorageClass{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "test-rsc",
+				},
+				Spec: v1alpha1.ReplicatedStorageClassSpec{
+					VolumeAccess: v1alpha1.VolumeAccessPreferablyLocal,
+				},
+			}
+		})
+
+		JustBeforeEach(func(ctx SpecContext) {
+			Expect(cl.Create(ctx, rsc)).To(Succeed(), "should create RSC")
+			Expect(cl.Create(ctx, rv)).To(Succeed(), "should create RV")
+		})
+
+		When("RV is being deleted", func() {
+			BeforeEach(func() {
+				now := metav1.Now()
+				rv.DeletionTimestamp = &now
+				rv.Finalizers = []string{"test-finalizer"}
+			})
```

Comment:
Крит:

```suggestion
			BeforeEach(func() {
				now := metav1.Now()
				rv.DeletionTimestamp = &now
				rv.Finalizers = []string{"test-finalizer"}
			})
			JustBeforeEach(func(ctx SpecContext){
				Expect(cl.Get(ctx, client.ObjectKeyFromObject(&rv), &rv)).To(Succeed())
				Expect(rv.DeletionTimestamp).ToNot(BeNil()) // тут сфэйлится
			})
```

Чтобы не сфэйлилось
```suggestion
			BeforeEach(func() {
				rv.Finalizers = []string{"test-finalizer"}
			})
			JustBeforeEach(func(ctx SpecContext){
				Expect(cl.Delete(ctx, rv)).To(Succeed())
				
				Expect(cl.Get(ctx, client.ObjectKeyFromObject(&rv), &rv)).To(Succeed())
				Expect(rv.DeletionTimestamp).ToNot(BeNil()) // тут норм
			})
```

Можно заносить в билдер сразу и не делать Delete. Слава умеет

Reply:
Блин, я проглядел. Я даже помню что вы эот разбирали, да.

Reply:
fixed